tareas/ErickGomez/Tarea1/Tarea-1.jmd

```diff
@@ -0,0 +1,169 @@
+
+En el mismo espíritu, uno puede definir al *par ordenado* $\overleftrightarrow{x} = (x, x^\prime) = x + \epsilon x^\prime$, con $x, x^\prime \in \mathbb R$, y donde $\epsilon$ lo definimos con la propiedad $\epsilon^2=0$. Es fácil pues convencerse que estas estructuras, que llamaremos *números duales*, cumplen:
+
+\begin{eqnarray}
+(x + \epsilon x^\prime) \pm (y + \epsilon y^\prime) & = & (x\pm y) + \epsilon (x^\prime\pm y^\prime),\\
+(x + \epsilon x^\prime) \cdot (y + \epsilon y^\prime) & = & (x\cdot y) + \epsilon (xy^\prime+y x^\prime).
+\end{eqnarray}
+
+Llamaremos a $x$ la *parte principal* y a $x^\prime$ la *parte derivada* de $\overleftrightarrow{x}$.
+
+## Ejercicio 1
+
+- A partir de $\overleftrightarrow{z} \cdot \overleftrightarrow{y} = \overleftrightarrow{x}$, obtener la expresión para $\overleftrightarrow{z}  = \overleftrightarrow{x} / \overleftrightarrow{y}$. Es decir, obtener qué es $z$ y $z^\prime$ en términos de las componentes de $\overleftrightarrow{x}$ y $\overleftrightarrow{y}$.
+Si es necesario pueden iniciar en $\overleftrightarrow{z} = {\overleftrightarrow{y}}^{-1}$
+
+Se tiene que
+$$
+\overleftrightarrow{z} \cdot \overleftrightarrow{y} =
+zy + \epsilon(zy' + z'y) = x + \epsilon x' = \overleftrightarrow{x}
+$$
+lo cual se cumple sii
+$$ zy = x \quad\text{y}\quad z' = zy' + z'y $$
+de donde 
+$$ z = \frac{x}{y}, $$
+luego:
+$$ x' = \left(\frac{x}{y}\right)y' + z'y 
+\quad\Rightarrow\quad 
+z' = \frac{yx' - xy'}{y^2} $$
+
+- De la expresión obtenida (y de las fórmulas anteriores), ¿qué podemos decir/concluir/interpretar de $z^\prime$?
+(En caso de falta de inspiración, ver [aquí](https://en.wikipedia.org/wiki/Automatic_differentiation#Automatic_differentiation_using_dual_numbers), pero es sugiero esforzarse y *arriesgarse* tratando de reconocer las expresiones.)
+
+Haciendo la analogía de la notación funcional de $f'$ como la derivada de $f$, si $z = \frac{x}{y}$, entonces $z'$ es la derivada de $z'$ dada por la regla del cociente de $\left(\frac{x}{y}\right)' = \frac{yx' - xy'}{y^2}$.
+
+## Ejercicio 2
+
+- Definir una estructura `Dual` que represente a los números duales; por sencillez pueden considerar que los campos del `Dual` son `Float64`, aunque pueden *osar* y tratar de implementar el caso paramétrico `Dual{T <: Real}`, donde `T` es el tipo de *ambos* campos.
+
+```julia
+Int64 == Real
+```
+
+```julia
+#"""
+#Definimos a un número Dual como una estructura mutable.
+#"""
+
+# \overleftrightarrow{z} = x + ϵy
+#mutable struct Dual{T <: Real, U <: Real}
+ #   x :: T
+  #  y :: U
+   # function Dual(x :: T, y :: U) where {T, U}
+    #    @assert length(x) == length(y) == 1
+     #   return new{T, U}(x,y)
+    #end
+#end
+```
+
+```julia
+# \overleftrightarrow{z} = x + ϵy
+mutable struct Dual{T <: Real}
+    x :: T
+    y :: T
+    function Dual(x :: T, y :: T) where {T}
+        @assert length(x) == length(y) == 1
+        return new{T}(x,y)
+    end
+end
+```
+
+```julia
+z = Dual(5.0,4)
```

Comment:
Esto da un error, dado que `x` y `y` tienen *ambos* que ser del mismo tipo; las siguientes celdas, entonces dan errores. Para evitar este error, es decir, imponer que `x` y `y` sean *ambos* del tipo `T` cuando no lo son, puedes usar [`promote`](https://docs.julialang.org/en/v1.3/manual/conversion-and-promotion/#).

Si usas `z = Dual(5.0, 2.2)` (de tal manera que ambos son del tipo `Float64`), todo funciona; aún así, es mejor corregir el error